### PR TITLE
podnetwork: Refactor to use the new type net/netip package

### PR DESCRIPTION
**Pull request by yoheiueda**
This is a follow-on PR to #947.

Go 1.18 introduces the new net/netip package to define a new IP address type netip.Addr. The new type is comparable, immutable, and lightweight compared to the old type net.IP.

Fixes https://github.com/confidential-containers/cloud-api-adaptor/issues/953

**References**

- https://tip.golang.org/doc/go1.18#netip
- https://djosephsen.github.io//posts/ipnet/

### CODE REVIEW

**Comment by bpradipt** [comment]
@yoheiueda do you want to target this for the 0.7.0 release ?

**Comment by yoheiueda** [comment]
@bpradipt Yes, I'd like to target this for 0.7.0.

I'll rebase my patch and make this PR ready for rewiew.

File: pkg/adaptor/server_test.go [hunk 1]
```diff
@@ -206,13 +206,9 @@ func (p *mockProvider) CreateInstance(ctx context.Context, podName, sandboxID st
...
 
-	ips := make([]net.IP, 2)
-	ips[0] = net.ParseIP(primaryIP)
-	ips[1] = net.ParseIP(secondaryIP)
-
-	if ips[0] == nil || ips[1] == nil {
-		return nil, fmt.Errorf("Could not parse IPs: primary IP %s and secondary IP %s", primaryIP, secondaryIP)
-	}
+	ips := make([]netip.Addr, 2)
+	ips[0] = netip.MustParseAddr(primaryIP)
```
Comment (huoqifeng) [inline pkg/adaptor/server_test.go:210]: Which will panic on error? so that the err check was deleted?

Comment (yoheiueda) [reply pkg/adaptor/server_test.go:210]: `netip.MustParseAddr(primaryIP)` will panic on errors.

https://pkg.go.dev/net/netip#MustParseAddr

I think this is OK in test code.

**Review by huoqifeng — APPROVED** [review]
/lgtm
I did not test it, but overall the replacing code looks good to me.

**Review by liudalibj — APPROVED** [review]
LGTM, thanks @yoheiueda 

I tested this PR.
- Build podvm image base the PR branch
```
/usr/bin/docker buildx build --build-arg ARCH=amd64 --build-arg CAA_SRC=https://github.com/yoheiueda/cloud-api-adaptor --build-arg CAA_SRC_REF=netip --build-arg BUILDER_IMG=quay.io/liudalibj/podvm-builder-ubuntu --file Dockerfile.podvm_binaries --platform linux/amd64 --provenance false --tag quay.io/liudalibj/podvm-binaries-ubuntu-amd64-yh:latest --push .

/usr/bin/docker buildx build --build-arg CLOUD_PROVIDER=generic --build-arg ARCH=amd64 --build-arg CAA_SRC=https://github.com/yoheiueda/cloud-api-adaptor --build-arg CAA_SRC_REF=netip --build-arg BINARIES_IMG=quay.io/liudalibj/podvm-binaries-ubuntu-amd64-yh --build-arg BUILDER_IMG=quay.io/liudalibj/podvm-builder-ubuntu --file Dockerfile.podvm --platform linux/amd64 --provenance false --tag quay.io/liudalibj/se-podvm-generic-ubuntu-amd64:netip . --load
```
- Create one new custom image 
```
./import.sh quay.io/liudalibj/se-podvm-generic-ubuntu-amd64:netip eu-gb --region eu-gb --bucket pp-e2e-test-images --instance SE-images-e2e-test-cos

Successfully uploaded 'podvm-d3af8b5-dirty-amd64.qcow2' to bucket 'pp-e2e-test-images'.
...
Created image podvm-d3af8b5-dirty-amd64 with id r018-c91bacfd-f86d-4b85-89fa-3456bae13525
``` 
- Build new CAA image
```
export registry=quay.io/liudalibj
make image
COMMIT=d3af8b5467c68c14ce877510ce7a9febdf13102e VERSION=unknown hack/build.sh
~/cloud-api-adaptor ~/cloud-api-adaptor
[+] Building 127.2s (20/20) FINISHED
 => [internal] load .dockerignore                                                                                                                                                                                                               0.0s
 => => transferring context: 267B                                                                                                                                                                                                               0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                                                            0.0s
 => => transferring dockerfile: 931B                                                                                                                                                                                                            0.0s
 => [internal] load metadata for docker.io/library/fedora:36                                                                                                                                                                                    0.6s
 => [internal] load metadata for quay.io/confidential-containers/golang-fedora:1.20.5-36                                                                                                                                                        0.4s
 => [builder-release 1/1] FROM quay.io/confidential-containers/golang-fedora:1.20.5-36@sha256:171e3674e14c01d9063985315ed601df8c2e99970aa8523ffa30b2c4bdf8e421                                                                                  0.1s
 => => resolve quay.io/confidential-containers/golang-fedora:1.20.5-36@sha256:171e3674e14c01d9063985315ed601df8c2e99970aa8523ffa30b2c4bdf8e421                                                                                                  0.1s
 => [base-release 1/1] FROM docker.io/library/fedora:36@sha256:64cd00a0e2b92d527c0a0954162a73e85f160e3a53c38325b51e87d6aab4e266                                                                                                                 0.0s
 => => resolve docker.io/library/fedora:36@sha256:64cd00a0e2b92d527c0a0954162a73e85f160e3a53c38325b51e87d6aab4e266                                                                                                                              0.0s
 => [internal] load build context                                                                                                                                                                                                               0.0s
 => => transferring context: 6.44kB                                                                                                                                                                                                             0.0s
 => CACHED [base-dev 1/1] RUN dnf install -y libvirt-libs genisoimage /usr/bin/ssh && dnf clean all                                                                                                                                             0.0s
 => CACHED [builder-dev 1/1] RUN dnf install -y libvirt-devel && dnf clean all                                                                                                                                                                  0.0s
 => CACHED [builder 1/8] WORKDIR /work                                                                                                                                                                                                          0.0s
 => CACHED [builder 2/8] COPY go.mod go.sum ./                                                                                                                                                                                                  0.0s
 => [builder 3/8] RUN go mod download                                                                                                                                                                                                          24.8s
 => [builder 4/8] COPY entrypoint.sh Makefile ./                                                                                                                                                                                                0.5s
 => [builder 5/8] COPY cmd   ./cmd                                                                                                                                                                                                              0.0s
 => [builder 6/8] COPY pkg   ./pkg                                                                                                                                                                                                              0.1s
 => [builder 7/8] COPY proto ./proto                                                                                                                                                                                                            0.0s
 => [builder 8/8] RUN make ARCH=amd64 COMMIT=d3af8b5467c68c14ce877510ce7a9febdf13102e VERSION=unknown RELEASE_BUILD=false cloud-api-adaptor                                                                                                    86.5s
 => [stage-5 1/1] COPY --from=builder /work/cloud-api-adaptor /work/entrypoint.sh /usr/local/bin/                                                                                                                                               1.9s
 => exporting to image                                                                                                                                                                                                                         12.2s
 => => exporting layers                                                                                                                                                                                                                         3.5s
 => => exporting manifest sha256:9346bc7cdc77be8fd5182cab1d087d70508a886768a656a4ac6fcf9052d80e63                                                                                                                                               0.0s
 => => exporting config sha256:c343e0f0f3d49f93a80f9d2864f316007df55582b81637449418913f4b0842c8                                                                                                                                                 0.0s
 => => exporting attestation manifest sha256:f5ac0dda28732b5cc973b5f53a3ca2ead20186ddc9b9c1d81214ba67ee27484e                                                                                                                                   0.0s
 => => exporting manifest list sha256:cf31bda49bb1b633a8c37ee79073ed8079eb03f11cb2c050a4aa369ab9beccd1                                                                                                                                          0.0s
 => => pushing layers                                                                                                                                                                                                                           0.4s
 => => pushing manifest for quay.io/liudalibj/cloud-api-adaptor:latest@sha256:cf31bda49bb1b633a8c37ee79073ed8079eb03f11cb2c050a4aa369ab9beccd1                                                                                                  0.9s
 => => pushing manifest for quay.io/liudalibj/cloud-api-adaptor:dev-d3af8b5467c68c14ce877510ce7a9febdf13102e@sha256:cf31bda49bb1b633a8c37ee79073ed8079eb03f11cb2c050a4aa369ab9beccd1                                                            0.5s
 => [auth] liudalibj/cloud-api-adaptor:pull,push token for quay.io                                                                                                                                                                              1.5s
~/cloud-api-adaptor
```
- Install CAA with the new built out podvm image and CAA image tag
```
root@liudali-csi-amd64-node-0:~/cloud-api-adaptor# kubectl describe po -n confidential-containers-system   cloud-api-adaptor-daemonset-6nsbh |grep image
  Normal  Pulling    9m     kubelet            Pulling image "quay.io/liudalibj/cloud-api-adaptor:latest"
  Normal  Pulled     8m59s  kubelet            Successfully pulled image "quay.io/liudalibj/cloud-api-adaptor:latest" in 385.861601ms (385.885101ms including waiting)
root@liudali-csi-amd64-node-0:~/cloud-api-adaptor#
```
- Run nginx demo
```
kubectl create -f ibmcloud/demo/nginx.yaml
...
root@liudali-csi-amd64-node-0:~/cloud-api-adaptor# kubectl get po
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   0          9m18s
root@liudali-csi-amd64-node-0:~/cloud-api-adaptor#
```
- Check CAA log
```
root@liudali-csi-amd64-node-0:~/cloud-api-adaptor# kubectl logs -n confidential-containers-system   cloud-api-adaptor-daemonset-6nsbh -f
$IBMCLOUD_API_KEY is SET
+ exec cloud-api-adaptor ibmcloud -iam-service-url https://iam.cloud.ibm.com/identity/token -vpc-service-url https://eu-gb.iaas.cloud.ibm.com/v1 -resource-group-id 0013399570e34af29d51788b600fc617 -key-id r018-0d1d29d6-df58-40e6-8576-9c630688e0d1 -image-id r018-c91bacfd-f86d-4b85-89fa-3456bae13525 -profile-name bx2-2x8 -zone-name eu-gb-2 -primary-subnet-id 0797-4d86be67-761c-4a24-9e32-54f8c46c0f51 -primary-security-group-id r018-57a4d4dd-67ff-49b5-afd2-43b9346dd930 -vpc-id r018-bd9aa88b-1d3c-4a9d-8ba4-7afd1f0c2302 -pods-dir /run/peerpod/pods -cri-runtime-endpoint /run/cri-runtime/containerd.sock -socket /run/peerpod/hypervisor.sock
cloud-api-adaptor version unknown
  commit: d3af8b5467c68c14ce877510ce7a9febdf13102e
  go: go1.20.5
cloud-api-adaptor: starting Cloud API Adaptor daemon for "ibmcloud"
2023/07/06 09:33:31 [adaptor/cloud/ibmcloud] ibmcloud-vpc config: ibmcloud.Config{ApiKey:"**********", IAMProfileID:"", CRTokenFileName:"/var/run/secrets/tokens/vault-token", IamServiceURL:"https://iam.cloud.ibm.com/identity/token", VpcServiceURL:"https://eu-gb.iaas.cloud.ibm.com/v1", ResourceGroupID:"0013399570e34af29d51788b600fc617", ProfileName:"bx2-2x8", ZoneName:"eu-gb-2", ImageID:"r018-c91bacfd-f86d-4b85-89fa-3456bae13525", PrimarySubnetID:"0797-4d86be67-761c-4a24-9e32-54f8c46c0f51", PrimarySecurityGroupID:"r018-57a4d4dd-67ff-49b5-afd2-43b9346dd930", SecondarySubnetID:"", SecondarySecurityGroupID:"", KeyID:"r018-0d1d29d6-df58-40e6-8576-9c630688e0d1", VpcID:"r018-bd9aa88b-1d3c-4a9d-8ba4-7afd1f0c2302"}
2023/07/06 09:33:31 [adaptor] server config: &adaptor.ServerConfig{TLSConfig:(*tlsutil.TLSConfig)(0xc0001ae300), SocketPath:"/run/peerpod/hypervisor.sock", CriSocketPath:"/run/cri-runtime/containerd.sock", PauseImage:"", PodsDir:"/run/peerpod/pods", ForwarderPort:"15150", ProxyTimeout:300000000000}
2023/07/06 09:33:31 [util/k8sops] initialized PeerPodService
2023/07/06 09:33:31 [adaptor] server started
2023/07/06 09:34:03 [podnetwork] routes on netns /var/run/netns/cni-e13fee4e-9c86-5f9b-a978-89af55033b4e
2023/07/06 09:34:03 [podnetwork]     default via 172.20.3.1 dev eth0
2023/07/06 09:34:03 [podnetwork]     172.20.0.0/16 via 172.20.3.1 dev eth0
2023/07/06 09:34:03 [adaptor/cloud] stored /run/peerpod/pods/54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea/daemon.json
2023/07/06 09:34:03 [adaptor/cloud] create a sandbox 54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea for pod nginx in namespace default (netns: /var/run/netns/cni-e13fee4e-9c86-5f9b-a978-89af55033b4e)
2023/07/06 09:34:03 [adaptor/cloud/ibmcloud] CreateInstance: name: "podvm-nginx-54103534"
2023/07/06 09:34:09 [adaptor/cloud/ibmcloud] podNodeIP[0]=10.242.64.6
2023/07/06 09:34:09 [adaptor/cloud] failed to create PeerPod: the server could not find the requested resource (post peerPods.confidentialcontainers.org)
2023/07/06 09:34:09 [adaptor/cloud] created an instance podvm-nginx-54103534 for sandbox 54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea
2023/07/06 09:34:09 [tunneler/vxlan] vxlan ppvxlan1 (remote 10.242.64.6:4789, id: 555000) created at /proc/1/task/12/ns/net
2023/07/06 09:34:09 [tunneler/vxlan] vxlan ppvxlan1 created at /proc/1/task/12/ns/net
2023/07/06 09:34:09 [tunneler/vxlan] vxlan ppvxlan1 is moved to /var/run/netns/cni-e13fee4e-9c86-5f9b-a978-89af55033b4e
2023/07/06 09:34:09 [tunneler/vxlan] Add tc redirect filters between eth0 and vxlan1 on pod network namespace /var/run/netns/cni-e13fee4e-9c86-5f9b-a978-89af55033b4e
2023/07/06 09:34:09 [adaptor/proxy] Listening on /run/peerpod/pods/54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea/agent.ttrpc
2023/07/06 09:34:09 [adaptor/proxy] established cri uds connection to /run/cri-runtime/containerd.sock
2023/07/06 09:34:09 [adaptor/proxy] Trying to establish agent proxy connection to 10.242.64.6:15150
2023/07/06 09:34:40 [adaptor/proxy] established agent proxy connection to 10.242.64.6:15150
2023/07/06 09:34:40 [adaptor/cloud] agent proxy is ready
2023/07/06 09:34:40 [adaptor/proxy] CreateSandbox: hostname:nginx sandboxId:54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea
2023/07/06 09:34:40 [adaptor/proxy]     storages:
2023/07/06 09:34:40 [adaptor/proxy]         mountpoint:/run/kata-containers/sandbox/shm source:shm fstype:tmpfs driver:ephemeral
2023/07/06 09:34:40 [adaptor/proxy] CreateContainer: containerID:54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea
2023/07/06 09:34:40 [adaptor/proxy]     mounts:
2023/07/06 09:34:40 [adaptor/proxy]         destination:/proc source:proc type:proc
2023/07/06 09:34:40 [adaptor/proxy]         destination:/dev source:tmpfs type:tmpfs
2023/07/06 09:34:40 [adaptor/proxy]         destination:/dev/pts source:devpts type:devpts
2023/07/06 09:34:40 [adaptor/proxy]         destination:/dev/shm source:/run/kata-containers/sandbox/shm type:bind
2023/07/06 09:34:40 [adaptor/proxy]         destination:/dev/mqueue source:mqueue type:mqueue
2023/07/06 09:34:40 [adaptor/proxy]         destination:/sys source:sysfs type:sysfs
2023/07/06 09:34:40 [adaptor/proxy]         destination:/dev/shm source:/run/kata-containers/sandbox/shm type:bind
2023/07/06 09:34:40 [adaptor/proxy]         destination:/etc/resolv.conf source:/run/kata-containers/shared/containers/54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea-99898951501915fd-resolv.conf type:bind
2023/07/06 09:34:40 [adaptor/proxy]     annotations:
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.sandbox-log-directory: /var/log/pods/default_nginx_891391fb-c943-455e-9e46-d3dce72fcdc2
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.container-type: sandbox
2023/07/06 09:34:40 [adaptor/proxy]         nerdctl/network-namespace: /var/run/netns/cni-e13fee4e-9c86-5f9b-a978-89af55033b4e
2023/07/06 09:34:40 [adaptor/proxy]         io.katacontainers.pkg.oci.bundle_path: /run/containerd/io.containerd.runtime.v2.task/k8s.io/54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.sandbox-cpu-shares: 2
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.sandbox-id: 54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.sandbox-cpu-quota: 0
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.sandbox-namespace: default
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.sandbox-name: nginx
2023/07/06 09:34:40 [adaptor/proxy]         io.katacontainers.pkg.oci.container_type: pod_sandbox
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.sandbox-memory: 0
2023/07/06 09:34:40 [adaptor/proxy]         io.kubernetes.cri.sandbox-cpu-period: 100000
2023/07/06 09:34:40 [adaptor/proxy] getImageName: no pause image specified uses default pause image: registry.k8s.io/pause:3.7
2023/07/06 09:34:40 [adaptor/proxy] CreateContainer: calling PullImage for "registry.k8s.io/pause:3.7" before CreateContainer (cid: "54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea")
2023/07/06 09:34:42 [adaptor/proxy] CreateContainer: successfully pulled image "registry.k8s.io/pause:3.7"
2023/07/06 09:34:42 [adaptor/proxy] StartContainer: containerID:54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea
2023/07/06 09:34:48 [adaptor/proxy] CreateContainer: containerID:a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43
2023/07/06 09:34:48 [adaptor/proxy]     mounts:
2023/07/06 09:34:48 [adaptor/proxy]         destination:/proc source:proc type:proc
2023/07/06 09:34:48 [adaptor/proxy]         destination:/dev source:tmpfs type:tmpfs
2023/07/06 09:34:48 [adaptor/proxy]         destination:/dev/pts source:devpts type:devpts
2023/07/06 09:34:48 [adaptor/proxy]         destination:/dev/mqueue source:mqueue type:mqueue
2023/07/06 09:34:48 [adaptor/proxy]         destination:/sys source:sysfs type:sysfs
2023/07/06 09:34:48 [adaptor/proxy]         destination:/sys/fs/cgroup source:cgroup type:cgroup
2023/07/06 09:34:48 [adaptor/proxy]         destination:/etc/config source:/run/kata-containers/shared/containers/a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43-3e385d5c5facdaef-config type:bind
2023/07/06 09:34:48 [adaptor/proxy]         destination:/etc/secret source:/run/kata-containers/shared/containers/a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43-5314acbcee7a8da1-secret type:bind
2023/07/06 09:34:48 [adaptor/proxy]         destination:/etc/hosts source:/run/kata-containers/shared/containers/a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43-c7a450b78cdb3bd8-hosts type:bind
2023/07/06 09:34:48 [adaptor/proxy]         destination:/dev/termination-log source:/run/kata-containers/shared/containers/a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43-fba21c260af7fad9-termination-log type:bind
2023/07/06 09:34:48 [adaptor/proxy]         destination:/etc/hostname source:/run/kata-containers/shared/containers/a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43-282c9939ae4cadac-hostname type:bind
2023/07/06 09:34:48 [adaptor/proxy]         destination:/etc/resolv.conf source:/run/kata-containers/shared/containers/a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43-f72a7be17123658b-resolv.conf type:bind
2023/07/06 09:34:48 [adaptor/proxy]         destination:/dev/shm source:/run/kata-containers/sandbox/shm type:bind
2023/07/06 09:34:48 [adaptor/proxy]         destination:/var/run/secrets/kubernetes.io/serviceaccount source:/run/kata-containers/shared/containers/a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43-ef719b42e5138b2f-serviceaccount type:bind
2023/07/06 09:34:48 [adaptor/proxy]     annotations:
2023/07/06 09:34:48 [adaptor/proxy]         io.kubernetes.cri.container-type: container
2023/07/06 09:34:48 [adaptor/proxy]         io.kubernetes.cri.container-name: nginx
2023/07/06 09:34:48 [adaptor/proxy]         io.katacontainers.pkg.oci.container_type: pod_container
2023/07/06 09:34:48 [adaptor/proxy]         io.katacontainers.pkg.oci.bundle_path: /run/containerd/io.containerd.runtime.v2.task/k8s.io/a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43
2023/07/06 09:34:48 [adaptor/proxy]         io.kubernetes.cri.sandbox-namespace: default
2023/07/06 09:34:48 [adaptor/proxy]         io.kubernetes.cri.sandbox-name: nginx
2023/07/06 09:34:48 [adaptor/proxy]         io.kubernetes.cri.sandbox-id: 54103534f935164936083bbdcf84fadd03894f5aa88c0c23767f260a94bd16ea
2023/07/06 09:34:48 [adaptor/proxy]         io.kubernetes.cri.image-name: docker.io/library/nginx:latest
2023/07/06 09:34:48 [adaptor/proxy] getImageName: got image from annotations: docker.io/library/nginx:latest
2023/07/06 09:34:48 [adaptor/proxy] CreateContainer: calling PullImage for "docker.io/library/nginx:latest" before CreateContainer (cid: "a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43")
2023/07/06 09:34:52 [adaptor/proxy] CreateContainer: successfully pulled image "docker.io/library/nginx:latest"
2023/07/06 09:34:52 [adaptor/proxy] StartContainer: containerID:a7364aa41637ab2a5e7271af90332ff1c4c90a31d30687b3b39cca4d055d7a43
```

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @yoheiueda 
I didn't get time to test it. Thanks to @liudalibj for testing it.

**Comment by yoheiueda** [comment]
Thank you very much for testing and reviewing this PR!

**Pull request merged**